### PR TITLE
Bridge driver to support multiple networks

### PR DESCRIPTION
**Pull request by aboch**
- Will add tests later

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

File: drivers/bridge/bridge.go [hunk 1]
```diff
@@ -198,11 +218,18 @@ func (d *driver) Config(option map[string]interface{}) error {
...
 
 func (d *driver) getNetwork(id types.UUID) (*bridgeNetwork, error) {
-	// Just a dummy function to return the only network managed by Bridge driver.
-	// But this API makes the caller code unchanged when we move to support multiple networks.
 	d.Lock()
 	defer d.Unlock()
-	return d.network, nil
+
+	if id == "" {
+		return nil, types.BadRequestErrorf("Invalid network id: %s", id)
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: Make all error strings start with lower case in this PR

File: drivers/bridge/bridge.go [hunk 2]
```diff
@@ -145,6 +145,26 @@ func (c *NetworkConfiguration) Validate() error {
...
+		return false
+	}
+
+	// Also empty, becasue only one network with empty name is allowed
+	if c.BridgeName == o.BridgeName {
+		return true
+	}
+
+	// They must be in different subnets
+	if c.AddressIPv4.Contains(o.AddressIPv4.IP) ||
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: Check if `AddressIPv4` is not nil

File: drivers/bridge/bridge.go [hunk 3]
```diff
@@ -245,35 +272,70 @@ func (d *driver) CreateNetwork(id types.UUID, option map[string]interface{}) err
...
+			if nw.config.BridgeName == config.BridgeName {
+				d.Unlock()
+				return types.ForbiddenErrorf("Conflicts with network %s (%s)", nw.id, nw.config.BridgeName)
+			}
+			// If this network config specifies the AddressIPv4, we need
+			// to make sure it does not conflict with any previously allocated
+			// bridges. This could not be completely caught by the config conflict
+			// check, because networks which config does not specify the AddressIPv4
+			// get their address and subnet selected by the driver (see electBridgeIPv4())
+			if nw.config.AddressIPv4 != nil {
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: even at this point the `AddressIPv4` has not been elected. It only happens in `setupBridgeIPv4` so what happens here is essentially the same as config conflict check above.

File: drivers/bridge/bridge.go [hunk 4]
```diff
@@ -298,10 +360,6 @@ func (d *driver) CreateNetwork(id types.UUID, option map[string]interface{}) err
 		// assigned an IPv6 link-local address.
 		{config.EnableIPv6, setupBridgeIPv6},
 
-		// We ensure that the bridge has the expectedIPv4 and IPv6 addresses in
-		// the case of a previously existing device.
-		{bridgeAlreadyExists, setupVerifyAndReconcile},
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: Why did you remove this? I believe this is still needed.

File: drivers/bridge/bridge.go [hunk 5]
```diff
@@ -145,6 +145,26 @@ func (c *NetworkConfiguration) Validate() error {
 	return nil
 }
 
+// Conflict check if two NetworkConfiguration objects overlap in the multinetwork
+func (c *NetworkConfiguration) Conflict(o *NetworkConfiguration) bool {
+	if o == nil {
+		return false
+	}
+
+	// Also empty, becasue only one network with empty name is allowed
```
Comment (mavenugo) [inline drivers/bridge/bridge.go]: There shouldnt be any network with a bridgename that is empty. That should be marked as an error.
Even for the default case, we must force the client to pass the value (say `docker0` in this case)

Comment (aboch) [inline drivers/bridge/bridge.go]: This is current behavior. Just trying to maintain it.

File: drivers/bridge/bridge.go (hunk 2 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: Thanks, will check.

File: drivers/bridge/bridge.go (hunk 1 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: I will.

File: drivers/bridge/bridge.go (hunk 3 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: nw is one of the previously programmed network, not the actual one.

File: drivers/bridge/bridge.go (hunk 4 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: I don't know how this happened. Thanks for catching it.

File: drivers/bridge/bridge.go (hunk 3 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: So there is still is a bug here, the if check should be on the current network being configured, not the already programmed ones. Thanks.

**Comment by aboch** [comment]
Ping @mrjana. Thanks for your comments. Took care of them. Please take a look when you get a chance.

Also, given multiple network support, automated and enlarged system provided pool of networks in bridge.go.

File: drivers/bridge/bridge.go [hunk 6]
```diff
@@ -393,7 +460,11 @@ func (d *driver) CreateEndpoint(nid, eid types.UUID, epInfo driverapi.EndpointIn
 
 	// Get the network handler and make sure it exists
 	d.Lock()
-	n := d.network
+	n, ok := d.networks[nid]
+	if !ok {
+		d.Unlock()
+		return types.NotFoundErrorf("Network %s does not exist", nid)
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: You still have strings starting with upper case

File: drivers/bridge/bridge.go [hunk 7]
```diff
@@ -574,7 +645,11 @@ func (d *driver) DeleteEndpoint(nid, eid types.UUID) error {
...
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: err string in upper case

File: drivers/bridge/bridge.go [hunk 8]
```diff
@@ -645,7 +720,11 @@ func (d *driver) DeleteEndpoint(nid, eid types.UUID) error {
 func (d *driver) EndpointOperInfo(nid, eid types.UUID) (map[string]interface{}, error) {
 	// Get the network handler and make sure it exists
 	d.Lock()
-	n := d.network
+	n, ok := d.networks[nid]
+	if !ok {
+		d.Unlock()
+		return nil, types.NotFoundErrorf("Network %s does not exist", nid)
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: err string in upper case

File: drivers/bridge/setup_ipv4.go [hunk 9]
```diff
@@ -20,27 +20,24 @@ func init() {
...
+	// 10.[0-255].42.1/16
+	for i := 0; i < 256; i++ {
+		bridgeNetworks = append(bridgeNetworks, &net.IPNet{IP: []byte{10, byte(i), 42, 1}, Mask: mask})
+	}
+	// 192.168.[42-44].1/24
+	mask[2] = 255
+	for i := 42; i < 45; i++ {
+		bridgeNetworks = append(bridgeNetworks, &net.IPNet{IP: []byte{192, 168, byte(i), 1}, Mask: mask})
+	}
+	for i := 0; i < 256; i++ {
```
Comment (mrjana) [inline drivers/bridge/setup_ipv4.go]: You are adding `10.[0-255]` network twice

File: drivers/bridge/bridge.go (hunk 6 above)
Comment (aboch) [inline drivers/bridge/bridge.go]: will change

File: drivers/bridge/setup_ipv4.go (hunk 9 above)
Comment (aboch) [inline drivers/bridge/setup_ipv4.go]: Thanks, will fix

File: drivers/bridge/bridge.go (hunk 7 above)
Comment (mrjana) [inline drivers/bridge/bridge.go]: You still have error string starting in upper case

Comment (aboch) [inline drivers/bridge/bridge.go]: forgot to git add the file...

File: drivers/bridge/bridge.go [hunk 10]
```diff
@@ -245,35 +273,70 @@ func (d *driver) CreateNetwork(id types.UUID, option map[string]interface{}) err
...
-
 	// Create or retrieve the bridge L3 interface
 	bridgeIface := newInterface(config)
 	network.bridge = bridgeIface
 
+	// Verify network does not conflict with previously configured networks
+	// on parameters that were chosen by the driver.
+	d.Lock()
+	for _, nw := range d.networks {
+		if nw.id != id {
```
Comment (mrjana) [inline drivers/bridge/bridge.go]: Just one more comment. Can we lose the indent here by doing this:
``` 
 if nw.id == id {
    continue
}
```
There levels if indentation makes the code not readable.

**Comment by aboch** [comment]
ping @mrjana . Pls take a look.

**Comment by mrjana** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM. I think this has some level of dependency with #199 and hence will like to get that merged first followed by this PR.

**Comment by mrjana** [comment]
@aboch needs a rebase

**Comment by aboch** [comment]
ping @mrjana @mavenugo . I rebased. Thanks.

**Comment by mavenugo** [comment]
thanks @aboch .
